rp2040-hal/src/pio.rs

```diff
@@ -480,19 +486,13 @@ impl<SM: ValidStateMachine> UninitStateMachine<SM> {
         });
     }
 
-    /// Execute the instruction immediately.
-    // Safety: The Send trait assumes this is the only write to sm_instr while uninitialized. The
-    // initialized `StateMachine` may also use this register. The `UnintStateMachine` is consumed
-    // by `PIOBuilder.build` to create `StateMachine`
-    fn exec_instruction(&mut self, instruction: u16) {
-        self.sm()
-            .sm_instr
-            .write(|w| unsafe { w.sm0_instr().bits(instruction) })
-    }
-
     fn sm(&self) -> &rp2040_pac::pio0::SM {
         unsafe { &*self.sm }
     }
+
+    fn pio(&self) -> &rp2040_pac::pio0::RegisterBlock {
+        unsafe { &*self.block }
```

Comment:
Yes, `fn sm(…)`, `fn pio(…)` and `fn register_block(…)` all serve the same purpose. Safety cannot be documented at that point because it depends on what operation is actually done on the registers.

having them all as `unsafe fn` seems fair to me.